File: .github/scripts/post_to_x.ts
import { TwitterApi } from 'twitter-api-v2';
import * as fs from 'fs';

interface GitHubUser {
  login: string;
}

interface GitHubPullRequest {
  title?: string;
  html_url?: string;
  user?: GitHubUser;
}

interface GitHubEvent {
  pull_request?: GitHubPullRequest;
}

// PR情報をGitHub Actionsのイベントから取得
const eventPath = process.env.GITHUB_EVENT_PATH;
let pr: GitHubPullRequest = {};
if (eventPath && fs.existsSync(eventPath)) {
  const event: GitHubEvent = JSON.parse(fs.readFileSync(eventPath, 'utf8'));
  pr = event.pull_request || {};
}

const prTitle = pr.title ?? 'PRタイトル';
const prUrl = pr.html_url ?? 'PRリンク';
const prUser = pr.user?.login ?? 'ユーザー名';

const text = `提案したPRがマージされました！ありがとうございました！\nタイトル: ${prTitle}\nリンク: ${prUrl}\n提案者: ${prUser}さん`;
// const text = `test`;

const client = new TwitterApi({
  appKey: process.env.TWITTER_API_KEY!,
  appSecret: process.env.TWITTER_API_SECRET!,
  accessToken: process.env.TWITTER_ACCESS_TOKEN!,
  accessSecret: process.env.TWITTER_ACCESS_TOKEN_SECRET!,
});

(async () => {
  try {
    const tweet = await client.v2.tweet(text);
    console.log('Success:', tweet);
  } catch (e: unknown) {
    if (e instanceof Error) {
      console.error('Error:', e.message);
    } else {
      console.error('Unknown error:', e);
    }
    process.exit(1);
  }
})(); 

File: .github/scripts/post_to_x.py
import os
import requests
import json

# GitHub Actionsのイベント情報を取得
event_path = os.environ.get('GITHUB_EVENT_PATH')
if event_path and os.path.exists(event_path):
    with open(event_path, 'r') as f:
        event = json.load(f)
else:
    event = {}

pr = event.get('pull_request', {})
pr_title = pr.get('title', 'PRタイトル')
pr_url = pr.get('html_url', 'PRリンク')
pr_user = pr.get('user', {}).get('login', 'ユーザー名')

text = f"PRがマージされました！\nタイトル: {pr_title}\nリンク: {pr_url}\n提案者: {pr_user}"

# X APIに投稿
headers = {
    "Authorization": f"Bearer {os.environ['X_BEARER_TOKEN']}",
    "Content-Type": "application/json"
}
payload = {
    "text": text
}
response = requests.post("https://api.twitter.com/2/tweets", headers=headers, json=payload)
print(response.status_code, response.text) 